### ISSUE TITLE
Attribute name can have characters that are not allowed

### ISSUE THREAD

**Issue by darutoko** [pugjs/pug#2471]
Template:
`a(hre>f='google.com') Google`
Result:
`<a hre>f="google.com">Google</a>`

**Comment by TimothyGu** [comment]
It's a really easy-to-debug issue, and it's not really our responsibility to tell people who don't know HTML how to use HTML.

**Comment by ForbesLindesay** [comment]
I disagree.  This seems like a simple case that we can catch to make pug easier to use.  Preventing completely arbitrary characters will also give us more room to extend the syntax in innovative ways in the future.

**Comment by darutoko** [comment]
Since this is reopened now, I would like to know how does statement "not really our responsibility to tell people who don't know HTML how to use HTML' correlates with at least 3(sic!) checks for invalid class name? You don't care about HTML but you really, really, really, do care about class names?

**Comment by TimothyGu** [comment]
> I would like to know how does statement "not really our responsibility to tell people who don't know HTML how to use HTML' correlates with at least 3(sic!) checks for invalid class name?

This smells like trolling, but in good faith I am going to answer it anyway.

I assume you mean [these checks](https://github.com/pugjs/pug-lexer/blob/ee66475976e310553137eaf299201c3ab6333ee8/index.js#L406-L414).

The answer is simple: the things are parsed differently. Without rigid rules of when the class name is ending, we don't know when to start parsing other things. A class name may be followed with another class name, or an id, or `(` for starting a set of attributes, or `&attributes`, or a `:` denoting block expansion, or <code>&nbsp;</code> and then text. In attributes, there are only four very simple ways an attribute name can end: `)`, `,`, `=`, and <code>&nbsp;</code>.

> This seems like a simple case that we can catch to make pug easier to use

After re-reading the [HTML spec](https://www.w3.org/TR/html-markup/syntax.html#attribute-name) on attribute names, I guess the rules are indeed fairly easy to implement. While at it, we might also want to fix the currently badly broken `div(afds"afd")`, which renders to `<div afds "afd="afds "afd"></div>`.

**Comment by darutoko** [comment]
> This smells like trolling
It is not trolling.
> I assume you mean these checks.
Yes.
> The answer is simple: the things are parsed differently. Without rigid rules of when the class name is ending, we don't know when to start parsing other things. 
Now I suspect you are trolling me. [https://github.com/pugjs/pug-lexer/blob/ee66475976e310553137eaf299201c3ab6333ee8/index.js#L400-L401](This) lines are responsible for checking if you got a valid class name and can start parsing other things. The lines you mentioned are just excessive hand holding for some one who have no clue what valid class name looks like.
>  In attributes, there are only four very simple ways an attribute name can end: `)`, `,`, `=`, and ` `.
I think you missed `!`, which is totally valid attribute name character for some one who is familiar with HTML.

**Comment by ForbesLindesay** [comment]
> really, really, really, do care about class names

I really, really, really, do care about all code!  On the subject of class names though, I believe we are being a little bit too strict.  The spec I read does not appear to be enforced by modern browsers, and it is therefore common practice to use class names that don't match our regexp.  I therefore think we should remove the restriction about starting with `--`.
